### PR TITLE
avoid random random numbers by avoiding the use of std::random_device

### PR DESCRIPTION
**Pull request by wds15**
#### Submission Checklist

- [X] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
This fixes issue #890 which reports that some tests fail occasionally. It turns out that we use random random numbers ... that is the random seed of used for the random generator in some tests were derived using the `stan::random_device` facility such that the random number generator stream changed from test to test. This happens to trigger once in a while test failures.

This PR now replaces the `std::random_device` generator with a mersenne twister seeded with seed 1. Thus we will get always the same tests whenever things are sufficiently stable. This should avoid occasionally breaking tests while we still use random numbers as needed.

#### Intended Effect:
Make tests reliable.

#### How to Verify:
Run the tests
```
test/unit/math/fwd/mat/fun/matrix_exp_test.cpp
test/unit/math/fwd/mat/fun/matrix_exp_pade_test.cpp
test/unit/math/mix/mat/fun/append_array_test.cpp
test/unit/math/rev/mat/fun/matrix_exp_test.cpp
test/unit/math/prim/mat/prob/lkj_cov_test.cpp
test/unit/math/prim/mat/fun/matrix_exp_test.cpp
test/unit/math/prim/mat/fun/matrix_exp_pade_test.cpp
```

#### Side Effects:
None.

#### Documentation:
NA

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bgoodri** [comment]
I am not sure I agree with this approach. The point of non-deterministic tests is to notice when tests are flaky or there are edge cases. I get that it is annoying when nondeterministic tests fail for an unrelated PR, but it would seem that the tolerance should be upped or the tests / implementation should be improved. The human reviewer ought to be able to recognize when that is the case.

**Comment by wds15** [comment]
The matrix tests do fail once in a while which is why @syclik suggested to avoid true randomness (as I understood him). We need to get our test pipeline into a stable state again.

Would it make you feel better if we introduce a STAN_TEST_SEED macro which holds some integer to be used as initializer for the rngs? Then you could at least change that easily.

The other solution would be to less strict with the tolerances for the matrix exponential, but then we would have to loosen them quite a bit. Right now they are at 1E-10, but the tests did spit out differences of up to 1E-6 if I recall right. Thus we would have to set it to 1E-5 or so.

As I say, an unreliable testing pipeline is something we need to resolve. We could also merge this and file an issue (if reviewer agrees)?

**Comment by bgoodri** [comment]
I think loosening tolerances is better than reducing randomness.

Using only deterministic tests or equivalently using a fixed seed reduces
the coverage of the tests because they only get evaluated at values that
are known to be good or bad. There is a whole movement toward fuzzing in
order to find bugs in code. It is good to know that the matrix exponential
can only be accurate to 1E-6 in some cases.

If someone's PR fails because of some unrelated non-deterministic test,
then we should re-run the tests, approve the PR, and work to improve the
test or the implementation of the thing that fails with some small
probability.

**Comment by bgoodri** [comment]
Would you also agree that Stan Math should have more unit testing of
mathematical identities using non-deterministic inputs?

Most of Stan Math is unit testing of a few deterministic results that are
not identities and checking for exceptions with bad input. For the former,
at least for doubles / ints, we are usually not adding much value beyond
what is already in the Boost or Eigen unit tests. Our checking the
gradients is necessary but would be better if it were based on identities.

**Comment by wds15** [comment]
@bgoodri yep, testing identities would be great, but probably needs some clever framework.

I am closing this PR at the loosened tolerances were merged

**Comment by bob-carpenter** [comment]
How about creating a set of random stress tests that we run intermittently?  My main concern is that we don't slow down the dev process, specifically in the case where tests fail for reasons unrelated to the change being made.

For these numerical random stress tests, I'd be OK lowering tolerances in cases where we don't have good numerical analyses (i.e., almost everywhere).

As @bgoodri points out, there is a lot of room for improved testing.  

Analytic identities carry the risk of duplicating derivation mistakes in the original implementations.  Finite diffs have the nice property of checking the derivatives of the actual `double`-based function against the autodiff. 

The `double` and `int` tests do multiple things.  For even the simple library functions like `exp`, they are testing instantiatability.  Lack of instantiatability, particularly for the `int` case when faced with templates causing ambiguity, has caused a lot of bugs in the past.  This is the point that @syclik is always stressing.  

For the rest of our functions which aren't simple library functions, they're testing that we're getting the right values (or that Boost is getting the right values in some direct cases).  So I don't want to get rid of them.

Bottom line is that I'm OK with this, but it's a big enough policy decision that we should run it by @syclik for the math lib.

**Comment by bob-carpenter** [comment]
Whatever else we do, I like this idea:

**Comment by syclik** [comment]
It sounds like we really need two sets of tests:
- unit tests that should pass all the time
- randomized tests that may fail occasionally.

The reason to separate them out is because having intermittent failures is
really a recipe for driving developers insane. It's actually the reason
that got us testing in the first place; having things break that aren't
related to your own branch really makes it really difficult to debug.

But... there's tremendous value there if we choose to follow up on the
failures. Are we going to follow up on the failures or are they just going
to burn cycles? @bgoodri, I'd be much more inclined testing random inputs
if you're willing to help create a procedure for following up. (You don't
have to do all the work, but it would help if you could tell us what we
could do to debug.)

**Comment by sakrejda** [comment]
@syclik I agree that we need different sets of tests but randomization may not be what we need.  The math functions mostly run fast enough individually that for functions with a small number of arguments we can run complete tables of values to identify regions where they fail.  For functions (like the matrix exponential) that have a huge input space randomization makes a lot of sense, although even there it's probably pretty well known what causes failures (condition number and such).  Should probably continue this conversation on discourse.

**Comment by wds15** [comment]
Interesting. I did not expect to trigger such a discussion... which should be continued on discourse, yes. Just to make sure everyone got this: Now we have reduced tolerances for matrix_exp stuff in stan-math and we also have a ticket to follow up on this.

**Comment by bob-carpenter** [comment]
Yes, got it.  Thanks for fixing things so the tests pass.
